src/infra/core/src/dependency_graph_service_inmem.rs

```diff
@@ -193,6 +194,100 @@ impl DependencyGraphService for DependencyGraphServiceInMemory {
             .await
     }
 
+    async fn get_all_upstream_dependencies(
+        &self,
+        dataset_ids: Vec<DatasetID>,
+    ) -> Result<DatasetIDStream, GetUpstreamDependenciesError> {
+        self.ensure_datasets_initially_scanned()
+            .await
+            .int_err()
+            .map_err(GetUpstreamDependenciesError::Internal)
+            .unwrap();
+
+        let state = self.state.lock().await;
```

Comment:
Since we are working on a cloned + reversed graph below, do we need to be holding the lock for the entire duration of the operation, or we could clone under the lock and release it?

Reply:
The lock has to stay, see the comment about `Reversed` adapter that does not clone the graph.